### ISSUE TITLE
The family cannot fly because nobody has ever visited the flight masters: give the module an errand that does

### ISSUE THREAD

**Issue by quadseven** [quadseven/mod-overseer#388]
## What was measured

The family cannot use the flight network because they have never learned the
nodes, and the module says so every time it declines a flight. Verbatim from the
dev realm:

```
'<name>' is sent to 'at:1:-753.596,-2212.78,14.197' 4086 yards away and could
fly node 37 to node 80, but has not discovered node 32 on that route - trying
the next node out, or walking
```

and, in the same second, the walk it took instead:

```
'<name>' is sent to '...' and is routed round it - 9 surveyed nodes, 7874 yards
of walking legs and 1478 waypoints, finishing 344 yards from the aim where the
ordinary step takes over. No leg of it is a flight, a boat or a portal, because
the party walks together (#316)

'<name>' is sent to '...' and 4 of its 8 legs cross ground the other side
guards, with no way round that is not farther. It walks it.
```

A 4086 yard trip became a 7874 yard walk with four legs across guarded ground,
for the want of one node.

## What those nodes are

Read out of `TaxiNodes.dbc` rather than from memory. This realm's
`taxinodes_dbc` table ships empty in the pinned core, because AzerothCore reads
DBCs from extracted files, so the ids were confirmed against two independent
public DBC-to-SQL dumps that agree byte for byte:

| node | name | map |
|---|---|---|
| 37 | Nijel's Point, Desolace | 1 (Kalimdor) |
| 32 | Theramore, Dustwallow Marsh | 1 (Kalimdor) |
| 80 | Ratchet, The Barrens | 1 (Kalimdor) |

Sources: `Kaev/AzerothcoreDBCToSQL` `TaxiNodes.sql`, and
`ForgedWoW/WrathForgedCore` `WrathForged.Database/sql/base/dbc/TaxiNodes.sql`.

So the missing hop is Theramore, on the only route the graph offers from Nijel's
Point to Ratchet, and Ratchet is the town the campaign is based in.

## The mechanism already exists, and is a no-op in practice

`DiscoverFlightPointOnArrival` in `src/mod_overseer.cpp` already does the hard
half. Discovering a flight point is not a proximity effect: it is answered by
the packet a client sends when it opens a flight master's window
(`WorldSession::HandleTaxiQueryAvailableNodes`, which calls
`SendLearnNewTaxiNode`). A bot has no client, so that function calls
`PlayerTaxi::SetTaximaskNode` directly and reads the mask back afterwards rather
than trusting it, exactly as the trainer path reads `HasSkill` back.

Three things stop it from ever mattering:

1. **Nothing ever sends a character to a flight master on purpose.** The
   function runs opportunistically, on any creature errand that happens to end
   next to one. Its own comment says the decision of which zone, which
   character and when belongs to whatever plans travel, and that planner has
   never asked for a flight master.
2. **A specific node cannot be aimed at.** `flight master` is already a
   `TravelRoles()` keyword and `ResolveTravelTarget` resolves it to the NEAREST
   spawn carrying `UNIT_NPC_FLAG_FLIGHTMASTER`. To learn node 32 a character
   has to stand at node 32, which is by definition not the nearest one.
3. **Arrival is asked at the travel tolerance, not at the counter.** The
   creature is found with `FindNearestCreature(entry, TRAVEL_ARRIVED_YARDS)` at
   twelve yards, and the node is then taken from
   `sObjectMgr->GetNearestTaxiNode` on the character's own position, with no
   check that the node is the one this flight master answers for.
   `ConsiderFlight` already makes exactly that check against
   `TRAVEL_FLIGHT_NODE_MATCH_YARDS`; the discovery path does not. This is the
   same fault #378 fixed at a counter, and it wants the same answer: ask the
   live creature through the gate the executor will use.

## Suggested shape

* Let an errand name a node: `flight master:<nodeId>` resolved through
  `sTaxiNodesStore.LookupEntry(nodeId)` to the flight master standing for it,
  with the same node-to-creature agreement `ConsiderFlight` already requires.
  The keyword table and `at:`/`trigger:` prefixes are the precedent.
* Hold the character at the counter while the transaction lands, the way #378
  does, and verify the learned node is the one that was asked for.
* Make the refusal actionable: `ConsiderFlight` already knows the exact node it
  is missing, so the line that reports it should carry the aim string that would
  go and get it.
* Choosing which character goes and when stays with whatever writes the errand
  column, which is where this module already says that decision belongs. What
  this issue asks for is the errand, not the scheduler.

## Why this is the binding constraint

quadseven/mod-overseer#360 removed the other gate: the leader is no longer
grounded merely because a follower is on foot, and a party that can all board
now flies together. That change cannot fire while the members hold too few
nodes to route anywhere, which is what the line above is reporting. The network
is there, the module wants to use it, and nobody has ever walked up to the
counter.

## Out of scope

Boats and any other moving transport. Flight paths do not cross the ocean in
this expansion and this issue is about taxi nodes only.

**Comment by quadseven** [comment]
The other gate is now open: #390 stops the leader being grounded merely because a follower is on foot, and flies the whole party when everybody can board. That change cannot fire while the members hold too few nodes to route anywhere, which is what this issue is about. This is the binding constraint of the two.

**Comment by quadseven** [comment]
## The errand this issue asks for is cheaper than it looks, and one obvious shortcut is a trap

Three findings from reading the resolve path for #398 and #400, all of which change what
this issue has to build.

### 1. A specific flight master can already be aimed at, today, with no new vocabulary

This issue's reason 2 says a specific node cannot be aimed at, because `flight master`
resolves to the NEAREST spawn carrying the flag. True. But `ResolveTravelTarget`
(`src/mod_overseer.cpp:9302`) accepts a **bare creature entry id** at `:9390-9396`, and a
flight master is in the travel index for the ordinary reason: `BuildTravelIndex` (`:9236`)
keeps any spawn whose effective npcflag intersects `TravelAnyFlag()`, and
`UNIT_NPC_FLAG_FLIGHTMASTER` is in that union via the `flight master` keyword. Bragok's
`creature.npcflag` is 0, so `data.npcflag ? data.npcflag : creatureTemplate->npcflag` takes
8192 off the template and indexes him.

So `travel_npc = '16227'` walks a character to the Ratchet flight master now. The
`flight master:<nodeId>` form in the suggested shape is still nicer to write and still
worth having, because it makes the node the source of truth rather than the creature. It is
just not a blocker.

### 2. Aiming at a flight master by COORDINATE can never learn the node

Worth stating loudly, because it is the obvious robust-looking workaround and it fails
silently and permanently.

`DiscoverFlightPointOnArrival` (`:10256-10259`) opens with

```cpp
if (!entry)
    return;   // an `at:`/`trigger:` aim names no creature to check
```

and the `at:` branch of the resolve sets `outEntry = 0` deliberately at `:9335`. So an
`at:1:-898.246,-3769.65,11.7932` aim walks the character to Bragok's feet, arrives, ends
the errand, and discovers nothing. No warning, because the function returns before the
`HasNpcFlag` check that would have produced one.

Any design that resolves the flight master itself and aims at a position has to carry the
entry through to the arrival handler, or it is inert.

### 3. The gap is three characters, not five

Reading the taximasks properly: node 80 is word 2 bit 15, value 32768. Two of the five hold
it already. One holds 64 in that word, which is a different node. Two hold 0. So the Ratchet
errand is owed to three characters, and the earlier line about one member being able to fly
node 37 to node 80 is consistent with that rather than in conflict with it.

## What is still unowned: which node is worth going to

This issue is explicit that choosing which character goes and when belongs to whatever
writes the errand column. Agreed, and it is out of scope for #400 too. But there is a
question in between that nobody owns, and it is a pure one:

**a node is only worth learning if it connects to somewhere the family wants to be.**

The shipped travel graph has 3780 nodes and 15039 links and is the right source for that.
Sketch of a pure function, offered here rather than as a new issue so it stays attached to
the mechanism:

```
RankFlightErrands(candidates, known, wanted, limits) -> ranked errands
```

- `candidates`: each safe flight master, as `{nodeId, entry, x, y, walkYards, dangerYards}`.
  Safe means not hostile to us, which is the same faction reading #400 now prices ground
  with. The counts are 19 safe and 18 hostile on Kalimdor, 19 and 13 on Eastern Kingdoms, so
  roughly half the network is not ours and a naive nearest-flight-master rule walks into the
  other half.
- `known`: the union of the family's taximasks, so a node two of them hold still counts as
  owed by the third.
- `wanted`: the places the family actually operates. The bind, the campaign target, the
  staging point. Not a guess: these are already columns this module reads.
- The score is **what it unlocks**, not what it costs: how many `wanted` places become
  reachable by taxi once this node is held, given `known` plus this one. A node that
  completes a route is worth vastly more than a node that adds a leaf, and this is exactly
  the Theramore case in the issue body: node 32 is worth having only because it is the hop
  that makes 37 to 80 flyable.
- The cost is `walkYards + dangerYards`, where `dangerYards` is `ScoreGroundDanger` from
  #400 over the walk. That is what makes "nearest" mean "nearest that will not kill them"
  without a second faction rule.

And the first errand it should emit is the free one. The whole family is bound 82 yards
from Bragok. Three of them do not hold Ratchet. That is three walks of 82 yards, on ground
with nothing on it that is not grey to them, to unlock the node the campaign is based in.

## One thing #398 changes about the refusal line

This issue asks that `ConsiderFlight`'s refusal carry the aim string that would go and get
the missing node. That is still the right ask, and after #398 the aim string it should carry
is an **entry id**, not a name and not a coordinate. A name does not resolve at all
(`:9409`) and a coordinate cannot discover (`:10258`).

**Comment by quadseven** [comment]
## Live proof: the walk works, the discovery is never reached

Tested on the realm just now with the party standing in Ratchet. This both
confirms and corrects the earlier finding that `travel_npc = '16227'` would get
a character its flight point with no code change.

**What works.** A bare decimal creature entry resolves and walks the character
all the way there. Measured distances to Bragok at the end of the walk:

    Bork 5y   Grog 5y   Grug 6y   Og 8y   Ugga 7y

That is well inside `TRAVEL_ARRIVED_YARDS` (12), and the module confirms
arrival in its own words:

    00:50:08 INFO 'Grog' reached '16227' (creature 16227) - errand done, releasing

**What does not.** The entry errand is dispatched down the TRAINER path, which
rejects a flight master and drops the errand in the same instant:

    00:50:08 WARN creature 16227 ('Bragok') will not train 'Grog' - it is either
             not a trainer at all or not one this character may use.
             Dropping the errand

`DiscoverFlightPointOnArrival` is therefore never reached, and the taximask is
unchanged after repeated arrivals. Before and after, word 2 of the mask, where
node 80 lives at bit 15, value 32768:

    Bork  2170 0 64        no node 80
    Grog  134219832 0 32768  HAS node 80
    Grug  43114 0 32768      HAS node 80
    Og    2058 0 0          no node 80
    Ugga  2234 0 0          no node 80

Bragok's flags are 8195, which includes `UNIT_NPC_FLAG_FLIGHTMASTER` (8192),
so the npcflag test that discovery performs would pass. The character never
gets there.

## So the fix is small and precisely located

On arrival at a creature errand, the flight-master check must run
independently of, and not behind, the trainer check. Today an entry errand has
exactly one meaning, "go and train", and any other reason to stand next to a
creature is discarded.

Worth considering while fixing it: the same shape presumably blocks every other
reason to walk to a specific creature, including a repair vendor and an
auctioneer, which is the same family of failure as #363 and the repair work.

## Acceptance

With the party in Ratchet, aiming a character that lacks node 80 at entry
16227 must leave word 2 of its taximask with bit 15 set, and the log must say
which node was learned from which creature. Three characters are owed Ratchet:
Bork, Og and Ugga.

**Comment by quadseven** [comment]
## The discovery mechanism already works. The trial was run on a character that already held the node.

Correcting the reading that the entry errand is dispatched down the trainer path and
rejected before anything can look at the creature. That is not what the code does, and the
two log lines quoted for it prove the opposite.

`TrainOnArrival` **returns `true`** on the "will not train" path:

`src/mod_overseer.cpp:10180-10187`

```cpp
Trainer::Trainer* trainer = sObjectMgr->GetTrainer(entry);
if (!trainer || !trainer->IsTrainerValidForPlayer(bot))
{
    LOG_WARN(... "will not train '{}' ... Dropping the errand" ...);
    ClearLearnAim(name);
    return true;                                   // :10186
}
```

`true` makes `!TrainOnArrival(...)` false, so the `else if` at `:15597` is not taken and
control falls into the `else`. The line `reached '{}' (creature {}) - errand done,
releasing` exists **exactly once** in this file, at `:15860`, and it sits three statements
after `DiscoverFlightPointOnArrival` at `:15857`. There is no path that prints it without
having called discovery first.

So discovery ran, reached the creature, and returned at:

`src/mod_overseer.cpp:10305-10306`

```cpp
if (bot->m_taxi.IsTaximaskNodeKnown(node))
    return;   // already known - the ordinary case after the first visit
```

silently, because the character used for the trial is **one of the two that already hold
node 80**. By the taximask reading in this thread, two of five hold it and three do not.

The arithmetic checks out, incidentally: `PlayerTaxi.h:35-40` masks on `(nodeidx - 1)`, so
node 80 is field `79 / 32` = 2, bit `79 % 32` = 15, value 32768. Exactly as reported.

**So the acceptance already passes for a character that lacks the node.** Aiming Bork, Og or
Ugga at entry 16227 should set the bit today. What the mechanism did not do is say anything
when it correctly did nothing, which is why a working no-op and a broken mechanism looked
identical for a night. #402 fixes that: discovery now speaks on every outcome after the
flight-master gate, including "already knows node N", and the two checks before that gate
stay silent because they fire on every vendor and trainer arrival in the game.

## What WAS genuinely blocking, and is worth knowing before the errand is scheduled

Walking a character to a flight master **cancelled its training plan.** `TrainOnArrival` runs
on every creature arrival for which a learn plan exists, with no regard for what the errand
was for, and on a non-trainer it calls `ClearLearnAim` at `:10185`. The learn aim is a
separate column from the travel aim.

An errand that visits flight masters on purpose would have destroyed a profession once per
trip. That is fixed in #402 as well, so this issue's errand is now safe to issue rather than
merely possible to issue.

## What is still owed to this issue

Unchanged by any of the above: choosing WHICH node is worth learning and which character
goes. The ranking sketch posted earlier stands, and the first errand it emits is still the
free one - three characters, 82 yards, on ground with nothing on it that is not grey to
them.
